state_machine_design.py

```diff
@@ -0,0 +1,355 @@
+        try:
+            with open(sm.ens_filename, "r", encoding="utf-8") as f_in:
+                sm.ens = json.load(f_in)
+
+            f_in.close()
+            # f = open(sm.ens_filename)
+            # sm.ens = json.load(f)
+            # f.close()
+            sm.ens_index = sm.ens["next_ensemble"]
+            sm.ens_list = sm.ens["ensemble_list"]
+        except:
+            sm.err_code = NO_ENS_FILE
+            sm.err_msg = "active_ensembles.json not found"
+
+        now = time.localtime()
+        curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+        print("waking up at: " + str(curr_time_seconds))
+
+    def update(self, sm):
+        print("WAKE_UP update func")
+        # if active ensembles file not found or other reading with error,
+        # transition to ERROR state
+        if sm.err_code != NO_ERR:
+            sm.state = ERROR()
+        else:
+            # otherwise always
+            # transition to CHECK_TIME state
+            sm.state = CHECK_TIME()
+
+
+class CHECK_TIME(State):
+    '''
+    CHECK_TIME determines if the machine should skip an ensemble,
+    run an ensemble, or sleep until the next ensemble.
+    '''
+
+    def __init__(self):
+        self.check_time_ctrl = SKIP
+
+    def process(self, sm):
+        print("CHECK_TIME process func")
+        print("ens index: " + str(sm.ens_index))
+
+        if sm.ens_index < len(sm.ens_list):
+            # read time from ensemble and compare to current_time
+            now = time.localtime()
+            curr_time_seconds = hmsToSeconds(now.tm_hour, now.tm_min, now.tm_sec)
+
+            nearest_ens_time = sm.ens_list[sm.ens_index]["start_time"]
+
+            if nearest_ens_time < curr_time_seconds:
+                print("time past current ens, skipping")
+                self.check_time_ctrl = SKIP
+                sm.err_code = MISSED_ENS
+                # sm.err_msg = "Current ensemble past time, skipping ensemble "
+                # + sm.ens_list[sm.ens_index]["title"] + " at " + curr_time_seconds
+            elif nearest_ens_time == curr_time_seconds:
+                print("correct time for ensemble: " + sm.ens_list[sm.ens_index]["title"])
+                self.check_time_ctrl = NOW
+            else:
+                print("ensemble " + sm.ens_list[sm.ens_index]["title"] +
+                      " is still in future, go to sleep")
+                self.check_time_ctrl = WAIT
+        else:
+            print("index beyond last ens, resetting")
+            self.check_time_ctrl = RESET
+            sm.ens_index = 0
+
+
+    def update(self, sm):
+        print("CHECK_TIME update func")
+
+        # if current_time is passed current_ensemble time,
+        # transition to ITERATE
+        if self.check_time_ctrl == SKIP:
+            # sm.state = ITERATE() # switch with below line to disable error log on skip
+            sm.state = ERROR() # do we want to log this as error then recover?
```

Comment:
I think using the error state partially as a logger could be good. For example, the missed ensemble error isn't worth stopping execution over but would be good to log. Might require some different logic for different errors though.

Reply:
The only error right now that will shut down execution is the missing ensembles file. Right now if we skip an ensemble, the ERROR state should print out the error and then redirect control so that it then checks the next ensemble in the list.

Reply:
That sounds good for now then. I might address logging in a future commit since we haven't really been doing it all along, but printing is fine for the time being.